dissect: refuse to use diskseq if ID_IGNORE_DISKSEQ=1 is specified

yuwata opened the pull request:
Moreover, even if the property is not specified, let's refuse to use diskseq for MD devices, as udevd may not be running yet.

Follow-up for 1a81ddef00a0a25f6bcdd1e6633430e8b240b87f and 8716a76ce2058876abc7c5b1001822ac9e08a617.

Fixes #32107.

YHNdnzj commented on src/shared/dissect-image.c:
```diff
@@ -515,6 +516,37 @@ static void dissected_partition_done(DissectedPartition *p) {
 
 #if HAVE_BLKID
+static int diskseq_should_be_used(
+                const char *whole_devname,
```
assert?

poettering commented:
So it certainly makes sense to check if ID_IGNORE_DISKSEQ=1 is set, and not do diskseq in that case. I am completely on board with that. We have the property after all, hence we should honour it.

But the explicit check for md sounds wrong to me. We should find a better approach. i.e. instead of disablng diskseq on md, we should find a way to disable gpt-auto as a whole on md. 

My understanding was that md doesn't have part scanning on anyway, and hence we shouldn't be tempted to use it. But maybe we actually never check for that?

Maybe we should just teach gpt-aut-generator to skip block devices where blockdev_partscan_enabled() doesn't return true?

yuwata commented:
Not sure, but we can create partition on MD device...

poettering commented:
> Not sure, but we can create partition on MD device...

Well, sure you can call fdisk on an MD device. But that doesn't mean the kernel picks it up, unless part scanning is actually enabled for the device. Question is: is part scanning enabled for MD devices by default? Can it be toggled from userspace?

Whether it is enabled can be read from the "capability" sysfs attr, it's a bit messy though, see what blockdev_partscan_enabled() does... 

I am a bit lazy to check if the flag is set for MD devices. A quick google search doesn't say anything about it.

yuwata commented:
```
[   18.366814] testsuite-64.sh[596]: + udevadm wait --settle --timeout=30 --removed /dev/md/mdmirpar /dev/disk/by-id/md-name-H:mdmirpar /dev/disk/by-id/md-uuid-aaaaaaaa:bbbbbbbb:cccccccc:00001010 /dev/disk/by-label/mdmirpar_part /dev/md3
[   18.371990] testsuite-64.sh[596]: + mdadm --assemble /dev/md/mdmirpar --name mdmirpar -v
[   18.376207] testsuite-64.sh[2996]: mdadm: looking for devices for /dev/md/mdmirpar
[   18.378404] testsuite-64.sh[2996]: mdadm: no recogniseable superblock on /dev/sdf4
[   18.380634] testsuite-64.sh[2996]: mdadm: no recogniseable superblock on /dev/sdf3
[   18.382557] testsuite-64.sh[2996]: mdadm: no recogniseable superblock on /dev/sdf2
[   18.384889] testsuite-64.sh[2996]: mdadm: Cannot assemble mbr metadata on /dev/sdf1
[   18.389368] testsuite-64.sh[2996]: mdadm: Cannot assemble mbr metadata on /dev/sdf
[   18.391211] testsuite-64.sh[2996]: mdadm: no recogniseable superblock on /dev/sde
[   18.394223] testsuite-64.sh[2996]: mdadm: /dev/sdd is identified as a member of /dev/md/mdmirpar, slot 3.
[   18.394416] testsuite-64.sh[2996]: mdadm: /dev/sdc is identified as a member of /dev/md/mdmirpar, slot 2.
[   18.394750] testsuite-64.sh[2996]: mdadm: /dev/sdb is identified as a member of /dev/md/mdmirpar, slot 1.
[   18.394750] testsuite-64.sh[2996]: mdadm: /dev/sda is identified as a member of /dev/md/mdmirpar, slot 0.
[   18.414462] testsuite-64.sh[2996]: mdadm: added /dev/sdb to /dev/md/mdmirpar as 1
[   18.422546] testsuite-64.sh[2996]: mdadm: added /dev/sdc to /dev/md/mdmirpar as 2
[   18.434379] testsuite-64.sh[2996]: mdadm: added /dev/sdd to /dev/md/mdmirpar as 3
[   18.443398] testsuite-64.sh[2996]: mdadm: added /dev/sda to /dev/md/mdmirpar as 0
[   18.450530] testsuite-64.sh[2996]: mdadm: /dev/md/mdmirpar has been started with 4 drives.
(snip)
[   21.550991] testsuite-64.sh[596]: + udevadm info --attribute-walk /dev/md/mdmirpar
[   21.569828] kernel: block md127: the capability attribute has been deprecated.
[   21.557053] testsuite-64.sh[3763]: Udevadm info starts with the device specified by the devpath and then
[   21.557053] testsuite-64.sh[3763]: walks up the chain of parent devices. It prints for every device
[   21.557053] testsuite-64.sh[3763]: found, all possible attributes in the udev rules key format.
[   21.557053] testsuite-64.sh[3763]: A rule to match, can be composed by the attributes of the device
[   21.557053] testsuite-64.sh[3763]: and the attributes from one single parent device.
[   21.557053] testsuite-64.sh[3763]:   looking at device '/devices/virtual/block/md127':
[   21.557053] testsuite-64.sh[3763]:     KERNEL=="md127"
[   21.557053] testsuite-64.sh[3763]:     SUBSYSTEM=="block"
[   21.557053] testsuite-64.sh[3763]:     DRIVER==""
[   21.557053] testsuite-64.sh[3763]:     ATTR{alignment_offset}=="0"
[   21.557053] testsuite-64.sh[3763]:     ATTR{capability}=="0"
(snip)
```
So, capability is zero, thus `GENHD_FL_NO_PART_SCAN` flag is not set.

But, interestingly(?), the kernel emits `block md127: the capability attribute has been deprecated.` warning message. So, I am not sure the capability `0` really means part scan is enabled...

yuwata commented:
Hmm. looking at `disk_capability_show()` in block/genhd.c, it always returns `0`... We need to update the logic...

yuwata commented:
The `GENHD_FL_NO_PART_SCAN` flag was introduced by https://github.com/torvalds/linux/commit/d27769ec3df1a8de9ca450d2dcd72d1ab259ba32 (v3.2). But the flag is also effectively implied when `minors` element of `struct gendisk` is `1`, which can be check with `ext_range` sysfs attribute. Explicit flag (`GENHD_FL_NO_PART_SCAN`) can be obtained from `capability` sysattr.

By https://github.com/torvalds/linux/commit/1ebe2e5f9d68e94c524aba876f27b945669a7879 (v5.17), we can check the flag from `ext_range` sysfs attribute directly.

As v5.17 is slightly new, we need to do the following:
1. first check `ext_range` sysattr, `1` means `GENHD_FL_NO_PART_SCAN` is set.
2. if `ext_range` is `> 1`, for older kernels, check `capability` sysattr.

yuwata commented:
@poettering `blockdev_partscan_enabled()` is fixed. Also, the questionable commit is split into two. I hope you like commits except for the last one.

poettering commented:
Hmm, so the kernel broke ABI here? Oh christ.

poettering commented:
> But, interestingly(?), the kernel emits `block md127: the capability attribute has been deprecated.` warning message. So, I am not sure the capability `0` really means part scan is enabled...

Oh man. Kernel broke API *big* *time* here. What a mess. Can't they use google or a frickin github code search to see if the *documented* APIs they provide are actually used by anyone before they deprecate them?

"We never break userspace", ha!

poettering commented:
I reported this to to linux-block btw. No idea if this will help in any way though.

https://lore.kernel.org/linux-block/ZhQJf8mzq_wipkBH@gardel-login/T/#u

yuwata commented:
> I reported this to to linux-block btw. No idea if this will help in any way though.
> 
> https://lore.kernel.org/linux-block/ZhQJf8mzq_wipkBH@gardel-login/T/#u

Thanks. But the sysattr was deprecated in v6.3, so it seems little bit late to complain about the change. I think we need anyway check both attributes.

poettering commented:
> Thanks. But the sysattr was deprecated in v6.3, so it seems little bit late to complain about the change. I think we need anyway check both attributes.

"deprecated"? A deprecation usually doesn't break things. The removal does. Their "deprecation" is more of an amputation, since it hardcodes the fields to zero.

yuwata commented:
> > Thanks. But the sysattr was deprecated in v6.3, so it seems little bit late to complain about the change. I think we need anyway check both attributes.
> 
> "deprecated"? A deprecation usually doesn't break things. The removal does. Their "deprecation" is more of an amputation, since it hardcodes the fields to zero.

Yeah, I completely agree with what you say. But anyway, that's too late, unfortunately. Though, I am not sure about the maintenance rules of the kernel, but I guess some of 6.x are already EOL, and that will never fixed on these branches...

poettering commented:
So I guess the "ext_range" thing should work, as I read the kernel sources. But "ext_range" is entirely undocumented afaics. If they willy-nilly change even documented sysattrs, this makes me a bit afraid of what they are going to do to the undocumented ones.

But I guess we have to work with what we get I guess, and make the best of the jello that the kernel API sometimes is. Hence fine by me.

poettering commented on src/shared/blockdev-util.c:
```diff
@@ -356,40 +357,56 @@ int lock_whole_block_device(dev_t devt, int operation) {
 
-        r = read_one_line_file(p, &buf);
-        if (r == -ENOENT) /* If the capability file doesn't exist then we are most likely looking at a
+        r = device_get_sysattr_int(dev, "ext_range", &ext_range);
```
why not make this `unsigned`? I mean, if the value is negative things are really broken, no?

yuwata replied:
Yeah, I thought so, but the kernel uses %d for this attribute.

poettering replied:
yeah, kernel devs are sometimes sloppy. I'd still claim we should refuse actually negative values (and zero) here though and report them as failures instead of treating them similar to 1

poettering commented:
This all is a lot worse than thought:

https://lore.kernel.org/linux-block/ZhRSVSmNmb_IjCCH@gardel-login/

I am not sure what to think about the Linux block maintainers now? 

Next time when the kernel devs come with their "we don't break userspace" mantra I'll slap them with this shit they pulled here left and right.

poettering commented:
hmm, so i guess we could try to issue LOOP_GET_STATUS on the device, which will give us the correct answer if the block device is a loop device.

The ext_range check is not going to work on partition block devices i guess (because the sysattr doesn't exist there at all, and doesn't contain 1 hence, even though part scanning for partition block devices is definitely off)

neilbrown commented:
It is not clear to me what md is doing wrong.  It doesn't explicitly handle diskseq at all.   All it does is provide a check_events function which sometimes returns DISK_EVENT_MEDIA_CHANGE.  It will only return this once for any change.  This is handled by (I think) by calls to disk_check_media_change() which md calls in md_open which is where other drivers call it.

Maybe we need:
	disk->events		= DISK_EVENT_MEDIA_CHANGE;
	disk->event_flags	= DISK_EVENT_FLAG_UEVENT;

Loop and sr have this, and sd has it for removable devices.
Do you have any notes anywhere on precisely what is going wrong with md??

poettering commented:
I have never used mdadm. But the original issue (https://github.com/systemd/systemd/issues/32107) is about something like this:

1. An early-boot component of systemd (`systemd-gpt-auto-generator`) looks at the block device behind the root fs at the moment of the initrd→host transition. It scans its GPT partition table, and figures out if there are partitions on it that shall be automatically mounted without involvement of /etc/fstab or suchlike. It then automatically generates systemd .mount units for them to mount them in the right places. The device nodes listed in the .mount units are using the /dev/disk/by-diskseq/* symlinks to the devices, i.e. that allow referencing device nodes in a stable way that doesn't suffer by media change attacks that much (at least if used properly). This works great for all block device types, except for MD. If the root fs is on MD it appears there's an extra media change generated between the initrd→host transition and late boot, so that the .mount units end up timing out, since the /dev/disk/by-diskseq/* symlinks never get created matching the diskseq that we initially saw.

Hence: what precisely is generating these media change events in MD spuriously, that causes the diskseq nr of the block device to change?

We basically want that if a block device is available during the initrd→host transition with diskseq nr X, then it sticks to X until it is disassembled again. But for some reason X changes to Y (with Y != X) during early boot, and that's a problem.

yuwata commented:
The first three commits are now in #32192 and #32193.

poettering commented:
I guess we can close this one, right?

iiuc #32192 should be sufficient for us not to try to dissecting MD devices anymore, because partscan should be off for them. Or did I get that wrong?

Hence we don't need any md-specific workaround their broken kernel code anymore, do we?

yuwata commented:
> I guess we can close this one, right?
> 
> iiuc #32192 should be sufficient for us not to try to dissecting MD devices anymore, because partscan should be off for them. Or did I get that wrong?
> 
> Hence we don't need any md-specific workaround their broken kernel code anymore, do we?

No, partscan seems enabled even on MD devices, as `ext_range` is positive (and `capabilities` is zero).
See https://github.com/systemd/systemd/pull/32111#issuecomment-2041724696.

I think using by-diskseq is an optional feature. Of course, filtering MD in the library is dirty. So, I do not stick to the current form. But I think we should support MD devices gracefully in some way.

poettering commented:
well, it's not a common scenario, I don't really feel the pressure here to commit a workaround. To me this is clearly a bug in MD, it triggers diskseq updates where it really shouldn't. They need to fix this, not us, and given that setups with GPT on top of MD are kinda exotic and don#t really make too much sense I think we can sit this one out, and should leave it to the MD folks to clean up their act?

poettering commented:
@neilbrown does md have an issue tracker somewhere? should we report the diskseq borkage on MD there? or is it sufficient that we CC'ed you here?
